pkg/chaosdaemon/tasks/chaosonprocess.go

```diff
@@ -0,0 +1,252 @@
+	}
+
+	cm.processMap[pid] = processTask
+	err = cm.commit(uid, pid)
+	if err != nil {
+		_ = cm.taskManager.RecoverTask(uid)
+		delete(cm.processMap, pid)
+		return errors.Wrapf(err, "update new process")
+	}
+	return nil
+}
+
+// Apply the task when the target pid of task is already be Created.
+// If it comes a UID injected , Apply will return ChaosErr.ErrDuplicateEntity.
+// If the Process has not been Created , Apply will return ChaosErr.NotFound("PID").
+func (cm ChaosOnProcessManager) Apply(uid UID, pid PID, config TaskToProcess) error {
+	err := cm.taskManager.AddTask(uid, NewTask(pid, config))
+	if err != nil {
+		return err
+	}
+	err = cm.commit(uid, pid)
+	if err != nil {
+		_ = cm.taskManager.RecoverTask(uid)
+		return err
+	}
+	return nil
+}
+
+// Recover the task, if there is no task on PID or recovering the last task on PID.
+// Recover in ChaosCanRecover will run, if runs failed it will just return the error.
+// If Recover will fail , but developer wants to clear it run : cm.ClearProcessChaos(pid, true).
+// If PID is already recovered successfully, Recover will return ChaosErr.NotFound("PID").
+// If UID is not Applied or Created or the target PID of UID is not the import pid,
+// Recover will return ChaosErr.NotFound("UID").
+func (cm ChaosOnProcessManager) Recover(uid UID, pid PID) error {
+	uIDs := cm.taskManager.GetWithPID(pid)
+	if len(uIDs) == 0 {
+		return ChaosErr.NotFound("PID")
+	}
+	if len(uIDs) == 1 {
+		if uIDs[0] != uid {
+			return ChaosErr.NotFound("UID")
+		}
+		err := cm.ClearProcessChaos(pid, false)
+		if err != nil {
+			return err
+		}
+		err = cm.taskManager.RecoverTask(uid)
+		if err != nil {
+			cm.logger.Error(err, "recover task with error")
+		}
+		return nil
+	}
+
+	err := cm.taskManager.RecoverTask(uid)
+	if err != nil {
+		cm.logger.Error(err, "recover task with error")
+		return nil
+	}
+
+	err = cm.commit(uIDs[0], pid)
+	if err != nil {
+		return errors.Wrapf(err, "update new process")
+	}
+	return nil
+}
+
+func (cm ChaosOnProcessManager) commit(uid UID, pid PID) error {
+	task, err := cm.taskManager.SumTask(uid)
+	if err != nil {
+		return errors.Wrapf(err, "unknown recovering in the taskMap, UID: %v", uid)
+	}
+	process, ok := cm.processMap[pid]
+	if !ok {
+		return errors.Wrapf(ChaosErr.NotFound("PID"), "PID : %d", pid)
+	}
+	tasker, ok := task.Data.(TaskToProcess)
+	if !ok {
+		return errors.New("task.Data here must implement TaskToProcess")
+	}
+	_ = tasker.Assign(process)
+	if err != nil {
+		return err
+	}
+	err = process.Inject(pid)
+	if err != nil {
+		return errors.Wrapf(err, "inject existing process PID : %d", pid)
+	}
+	return nil
+}
+
+func (cm ChaosOnProcessManager) ClearProcessChaos(pid PID, ignoreRecoverErr bool) error {
+	if process, ok := cm.processMap[pid]; ok {
+		pRecover, ok := process.(ChaosCanRecover)
+		if !ok {
+			return errors.Wrapf(ChaosErr.NotImplemented("ChaosCanRecover"), "process")
+		}
+		err := pRecover.Recover(pid)
+		if err != nil && !ignoreRecoverErr {
```

Comment:
how about adding a log when `ignoreRecoverErr` is true?

Reply:
Fixed.